xc/device.ml

```diff
@@ -1430,6 +1431,110 @@ let can_surprise_remove ~xs (x: device) = Generic.can_surprise_remove ~xs x
+      let add c domid =
+        Hashtbl.replace ftod (Qmp_protocol.to_fd c) domid;
+        Hashtbl.replace dtoc domid c
+      let remove c domid =
+        Hashtbl.remove ftod (Qmp_protocol.to_fd c);
+        Hashtbl.remove dtoc domid
+      let domid_of fd = try Some (Hashtbl.find ftod fd) with Not_found -> None
+      let channel_of domid = try Some (Hashtbl.find dtoc domid) with Not_found -> None
+    end
+
+    module Epoll = Core.Linux_ext.Epoll
+    module Flags = Core.Linux_ext.Epoll.Flags
+
+    module Monitor = struct
+      let create () = (Core.Std.Or_error.ok_exn Epoll.create) ~num_file_descrs:1001 ~max_ready_events:1001
+      let add m fd = Epoll.set m fd Flags.in_
+      let remove m fd = Epoll.remove m fd
+      let wait m = Epoll.wait m ~timeout:`Never
+      let with_event ~ok = function | `Ok -> ok() | `Timeout -> debug "Shouldn't receive epoll timeout event in qmp_event_thread"
+    end
+    let m = Monitor.create ()
+
+    let monitor_path domid = Printf.sprintf "/var/run/xen/qmp-event-%d" domid
+    let debug_exn msg e = debug "%s: %s" msg (Printexc.to_string e)
+
+    let remove domid =
+      try
+        match Lookup.channel_of domid with
+        | None -> ()
+        | Some c -> Lookup.remove c domid;
+          Monitor.remove m (Qmp_protocol.to_fd c);
+          Qmp_protocol.close c;
+          debug "Removed QMP Event fd for domain %d" domid
+      with e -> debug_exn (Printf.sprintf "Got exception trying remove qmp on domain-%d" domid) e
+
+    let add domid =
+      try
+        match Lookup.channel_of domid with
+        | Some _ -> ()
+        | None -> let c = Qmp_protocol.connect (monitor_path domid) in
+          Qmp_protocol.negotiate c;
+          Lookup.add c domid;
+          Monitor.add m (Qmp_protocol.to_fd c);
+          debug "Added QMP Event fd for domain %d" domid
+      with e ->
+        debug_exn (Printf.sprintf "QMP domain-%d: negotiation failed: removing socket" domid) e;
+        remove domid
+
+    let wakeup domid =
+      add domid;
+      (* write single byte to wake up Monitor.wait *)
+      if Unix.write pipe_w " " 0 1 <> 1 then
+        debug "Pipe wrote error, failed to wake up qmp event thread"
+
+    let qmp_event_handle domid qmp_event =
+      (* This function will be extended to handle qmp events *)
+      debug "Got QMP event, domain-%d: %s" domid qmp_event.event
+
+    let qmp_event_thread () =
+      Monitor.add m pipe_r;
+      while true do
+        try
+          Monitor.wait m |> Monitor.with_event
+            ~ok:(fun () ->
+                Epoll.iter_ready m ~f:(fun fd flags ->
+                    match fd with
+                    | pipe when pipe = pipe_r ->
+                      if flags = Flags.in_ then ()
+                      else debug "Received unexpected epoll event on pipe_r in qmp_event_thread"
+                    | sock ->
+                      match Lookup.domid_of sock with
+                      | None -> ()
+                      | Some domid ->
```

Comment:
I did not enter too much in the details of the implementation.
From a brief look, you can reduce the nesting slightly and improve readability here with

```ocaml
| Some domid when flags = Flags.in_ -> ...
| _ -> debug "EPOLL error on domain-%d, close qmp socket" domid;
       remove domid
```

Reply:
You could also streamline the code slightly using the monadic option bind from the core library but maybe that's too much

Reply:
@mseri -- Thanks for your comments.
Seems `lib/option.ml` is not included in `core-113.33.03` which we are using but available in old versions like `core-109.12.00`

Now I use `let (>>=) m f = match m with | Some x -> f x | None -> ()`  to simplify the code.
Are you happy to review again?

Reply:
Are you sure? According to their documentation it's there: https://ocaml.janestreet.com/ocaml-core/113.33/doc/core/Std/Option.mod/